src/main/java/duke/task/Deadline.java

```diff
@@ -0,0 +1,59 @@
+package duke.task;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.regex.Pattern;
+
+/**
+ * Encapsulates a Deadline
+ */
+public class Deadline extends Task {
+    private String by;
+
+    /**
+     * Constructor for a Deadline
+     *
+     * @param name Name / Description of the deadline
+     * @param by Description of when the deadline task is to be completed by
+     */
+    public Deadline(String name, String by) {
+        super(name, 'D');
+        this.by = parseDateTime(by);
+    }
+
+    private String parseDateTime(String originalString) {
+        StringBuffer deadline = new StringBuffer(originalString.replace("/", "-"));
+        if (Pattern.compile("([0-9]{4})[-]([1][0-2]|0[1-9])[-](3[01]|[12][0-9]|0[1-9])\\s[0-2][0-9][0-6][0-9]")
+                .matcher(deadline).matches()) { //yyyy-mm-dd HHmm
+            deadline.insert(deadline.length() - 2, ":");
+            deadline.replace(deadline.indexOf(" "), deadline.indexOf(" ") + 1, "T");
+            return LocalDateTime.parse(deadline).format(DateTimeFormatter.ofPattern("dd MMM yyyy HH:mm"));
+        } else if (Pattern.compile("(3[01]|[12][0-9]|0[1-9])[-]([1][0-2]|0[1-9])[-][0-9]{4}\\s[0-2][0-9][0-6][0-9]")
+                .matcher(deadline).matches()) { //dd-mm-yyyy HHmm
+            deadline.insert(deadline.length() - 2, ":");
+            deadline.replace(deadline.indexOf(" "), deadline.indexOf(" ") + 1, "T");
+            String dd = deadline.substring(0, 2);
+            deadline.replace(0, 2, deadline.substring(6, 10)).replace(8, 12, dd);
+            return LocalDateTime.parse(deadline).format(DateTimeFormatter.ofPattern("dd MMM yyyy HH:mm"));
+        } else if (Pattern.compile("([0-9]{4})[-]([1][0-2]|0[1-9])[-](3[01]|[12][0-9]|0[1-9])")
+                .matcher(deadline).matches()) { //yyyy-mm-dd
+            return LocalDate.parse(deadline).format(DateTimeFormatter.ofPattern("dd MMM yyyy"));
+        } else if (Pattern.compile("(3[01]|[12][0-9]|0[1-9])[-]([1][0-2]|0[1-9])[-][0-9]{4}")
+                .matcher(deadline).matches()) { //dd-mm-yyyy
+            String dd = deadline.substring(0, 2);
+            deadline.replace(0, 2, deadline.substring(6, 10)).replace(8, 12, dd);
+            return LocalDate.parse(deadline).format(DateTimeFormatter.ofPattern("dd MMM yyyy"));
+        }
```

Comment:
This section may be too complicated (especially the if statements). Would it help if you perhaps initialize the patterns as individual variables first (with names to indicate the pattern used) and then use it in the conditions?